src/sss_client/ssh/sss_ssh_knownhosts.c

```diff
@@ -0,0 +1,211 @@
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "sss_ssh_get_ent() failed (%d): %s\n", ret, sss_strerror(ret));
+        goto done;
+    }
+
+    keyhost = talloc_asprintf(tmp_ctx, "%s,%s", canonname, host);
+    if (keyhost == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to allocate keyhost\n");
+        ERROR("Not enough memory\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    *_keyhost = talloc_steal(mem_ctx, keyhost);
+    *_ent = talloc_steal(mem_ctx, ent);
+    ret = EOK;
+
+done:
+    if (ai != NULL) {
+        freeaddrinfo(ai);
+    }
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+int main(int argc, const char **argv)
+{
+    TALLOC_CTX *mem_ctx = NULL;
+    int pc_debug = SSSDBG_TOOLS_DEFAULT;
+    const char *pc_domain = NULL;
+    const char *pc_target = NULL;
+    const char *keyhost;
+    struct poptOption long_options[] = {
+        POPT_AUTOHELP
+        { "debug", '\0', POPT_ARG_INT | POPT_ARGFLAG_DOC_HIDDEN, &pc_debug, 0,
+          _("The debug level to run with"), NULL },
+        { "domain", 'd', POPT_ARG_STRING, &pc_domain, 0,
+          _("The SSSD domain to use"), NULL },
+        POPT_TABLEEND
+    };
+    poptContext pc = NULL;
+    struct sss_ssh_ent *ent;
+    size_t i;
+    int ret;
+    errno_t res;
+
+    debug_prg_name = argv[0];
+
+    ret = set_locale();
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "set_locale() failed (%d): %s\n", ret, strerror(ret));
+        ERROR("Error setting the locale\n");
+        ret = EXIT_FAILURE;
+        goto fini;
+    }
+
+    mem_ctx = talloc_new(NULL);
+    if (!mem_ctx) {
+        ERROR("Not enough memory\n");
+        ret = EXIT_FAILURE;
+        goto fini;
+    }
+
+    /* parse parameters */
+    pc = poptGetContext(NULL, argc, argv, long_options, 0);
+    poptSetOtherOptionHelp(pc, _("HOST"));
+    while ((ret = poptGetNextOpt(pc)) > 0)
+        ;
+
+    DEBUG_CLI_INIT(pc_debug);
+
+    if (ret != -1) {
+        BAD_POPT_PARAMS(pc, poptStrerror(ret), ret, fini);
+    }
+
+    pc_target = poptGetArg(pc);
+    if (pc_target == NULL) {
+        BAD_POPT_PARAMS(pc, _("Host not specified\n"), ret, fini);
+    }
+
+    /* look up the public keys */
+    res = known_hosts(mem_ctx, pc_domain, pc_target, &ent, &keyhost);
+    if (res != EOK) {
+        // On a successful execution, even if no key was found,
+        // ssh expects EXIT_SUCCESS.
+        ret = (res == ENOENT ? EXIT_SUCCESS : EXIT_FAILURE);
+        goto fini;
+    }
+
+    /* If the other side closes its end of the pipe, we don't want this tool
+     * to exit abruptly, but to finish gracefully instead because the valid
+     * key can be present in the data already written
+     */
+    signal(SIGPIPE, SIG_IGN);
+
+    /* print results */
+    for (i = 0; i < ent->num_pubkeys; i++) {
+        ret = sss_ssh_print_pubkey(&ent->pubkeys[i], keyhost);
+        if (ret != EOK && ret != EINVAL) {
```

Comment:
Why is there exception for `EINVAL`?

Reply:
As far as I could see, `EINVAL` means the current key is invalid. There is nothing else to do in this case, but we want to print the remaing keys, as a best effort situation. `sss_ssh_authorizedkeys.c` is doing the same thing (I actually took it from there).

If you prefer, I could log a message when this error code happens.

Reply:
I don't see `sss_ssh_print_pubkey()` returning `EINVAL` explicitly.
So this means failed atomic-write..?

> sss_ssh_authorizedkeys.c is doing the same 

that doesn't necessarily mean it's correct :)

Reply:
> I don't see `sss_ssh_print_pubkey()` returning `EINVAL` explicitly. So this means failed atomic-write..?

`sss_ssh_get_pubkey_algorithm()` is returning this error code.

> > sss_ssh_authorizedkeys.c is doing the same
> 
> that doesn't necessarily mean it's correct :)

Right. I doesn't mean that it is correct. But I think it is correct and the author of that code agrees. :)